controller/konnect/reconciler_credential_secrets.go

```diff
@@ -0,0 +1,418 @@
+				Name:      secretName,
+				Namespace: consumer.Namespace,
+			},
+		})
+	}
+	return ret
+}
+
+// secretIsUsedByConsumerAttachedToKonnectControlPlane returns true if the Secret
+// is used by KongConsumer which refers to a KonnectGatewayControlPlane.
+func secretIsUsedByConsumerAttachedToKonnectControlPlane(cl client.Client) func(obj client.Object) bool {
+	return func(obj client.Object) bool {
+		secret, ok := obj.(*corev1.Secret)
+		if !ok {
+			ctrllog.FromContext(context.Background()).Error(
+				operatorerrors.ErrUnexpectedObject,
+				"failed to run predicate function",
+				"expected", "Secret", "found", reflect.TypeOf(obj),
+			)
+			return false
+		}
+
+		// List consumers using this Secret as credential.
+		kongConsumerList := configurationv1.KongConsumerList{}
+		err := cl.List(
+			context.Background(),
+			&kongConsumerList,
+			client.MatchingFields{
+				IndexFieldKongConsumerReferencesSecrets: secret.GetName(),
+			},
+		)
+		if err != nil {
+			return false
+		}
+
+		for _, kongConsumer := range kongConsumerList.Items {
+			cpRef := kongConsumer.Spec.ControlPlaneRef
+			if cpRef != nil &&
+				cpRef.Type == configurationv1alpha1.ControlPlaneRefKonnectNamespacedRef {
+				return true
+			}
+		}
+		return false
+	}
+}
+
+// Reconcile reconciles a Secrets that are used as Consumers credentials.
+func (r *KongCredentialSecretReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	var secret corev1.Secret
+	if err := r.client.Get(ctx, req.NamespacedName, &secret); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	const entityTypeName = "Secret"
+	logger := log.GetLogger(ctx, entityTypeName, r.developmentMode)
+	log.Debug(logger, "reconciling")
+
+	if !secret.GetDeletionTimestamp().IsZero() {
+		log.Debug(logger, "secret is being deleted")
+		return ctrl.Result{}, nil
+	}
+
+	cl := client.NewNamespacedClient(r.client, secret.Namespace)
+
+	credType, err := extractKongCredentialType(&secret)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Validate the Secret using the credential type label.
+	if err := validateSecret(&secret, credType); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// List consumers using this secret as credential.
+	kongConsumerList := configurationv1.KongConsumerList{}
+	err = cl.List(
+		ctx,
+		&kongConsumerList,
+		client.MatchingFields{
+			IndexFieldKongConsumerReferencesSecrets: secret.GetName(),
+		},
+	)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed listing KongConsumers for Secret: %w", err)
+	}
+
+	// If there are no Consumers that use the Secret then remove all the Credentials that use it.
+	if len(kongConsumerList.Items) == 0 {
+		if err := cl.DeleteAllOf(ctx, &corev1.Secret{},
+			client.MatchingFields{
+				IndexFieldKongConsumerReferencesSecrets: secret.Name,
+			},
+		); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed deleting Secret: %w", err)
+		}
+		return ctrl.Result{}, nil
+	}
```

Comment:
I don't quite understand this one: secrets are created by users, why are we deleting them in case no consumers reference them?

Reply:
Good catch. I must have rushed it. 9eb218021b1548ae2c005ac1da0e048657f244b1 should fix that.